joining/src/main/scala/akka/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala

```diff
@@ -0,0 +1,230 @@
+  import context.dispatcher
+
+  val cluster = Cluster(context.system)
+  private val TimerKeyResolveDNS = "resolve-dns-key"
+
+  //  // FIXME it's either buggy or has some assumptions that we do not want
+//  import com.lightbend.dns.locator.{ ServiceLocator ⇒ DnsServiceLocator }
+//  val dnsLocator: ActorRef = context.actorOf(DnsServiceLocator.props, "dnsServiceLocator")
+
+  private val dns = IO(Dns)(context.system)
+
+  private var lastContactsObservation: DnsServiceContactsObservation =
+    DnsServiceContactsObservation(Long.MaxValue, Nil)
+
+  /** Awaiting initial */
+  override def receive: Receive = {
+    case InitiateBootstraping ⇒
+      val serviceName = settings.effectiveServiceName(context.system)
+
+      log.info("Locating service members, via DNS lookup: {}", serviceName)
+      resolve(serviceName, settings.dnsResolveTimeout).pipeTo(self)
+
+      context become bootstraping(serviceName, sender())
+  }
+
+  /** In process of searching for seed-nodes */
+  def bootstraping(serviceName: String, replyTo: ActorRef): Receive = {
+    case Internal.AttemptResolve(name, timeout) ⇒
+      resolve(name, timeout).pipeTo(self)
+
+    case SrvResolved(name, records) ⇒
+      val recordsSize = records.size
+      log.info("Discovered {} [{}] observation: {}", recordsSize, name, records.mkString("[", ",", "]"))
+
+      onSrvResolved(serviceName, records, recordsSize)
+
+    case ex: Failure ⇒
+      log.warning("Resolve attempt failed! Cause: {}", ex.cause)
+      scheduleNextResolve(serviceName, settings.dnsResolveTimeout)
+
+    case ObtainedHttpSeedNodesObservation(infoFromAddress, observedSeedNodes) ⇒
+      log.info("Contact point [{}] returned [{}] seed nodes [{}], initiating cluster joining...", infoFromAddress,
+        observedSeedNodes.size, observedSeedNodes)
+
+      replyTo ! "JOINING SEED NODES!" // FIXME nicer typed message
+
+      val seedNodesList = observedSeedNodes.toList
+      cluster.joinSeedNodes(seedNodesList)
+
+      stopProbingContactPoints()
+      cluster.subscribe(self, ClusterEvent.InitialStateAsSnapshot, classOf[ClusterDomainEvent])
+      context become joining(seedNodesList)
+
+    case NoSeedNodesObtainedWithinDeadline() ⇒
+    // this means that, likely, no cluster exists in this deployment yet, the lowest addressed node should join itself.
+
+    // TODO as it is right now, this just means that one of the contact points was stable and did not return any seed nodes
+    // TODO this is likely fine, but we could add additional guards here, like:
+    // TODO  - like the dns IPs we get back also being stable for the stableMargin
+    // TODO  - or all children we spawned having reported that their deadlines passed
+    // this moment by definition "risky" though, so a matter of how much we think that would help
+  }
+
+  /** Awaiting joining to complete before shutting down the bootstrap actors */
+  def joining(seedNodes: immutable.Seq[Address]): Receive = {
+    case initialState: CurrentClusterState ⇒
```

Comment:
This will be sent immediately after subscribe, with current state, which might be before it has joined. Not sure what the purpose of this subscription is?
You could subscribe to `MemberJoined` but I don't see much point in waiting for that. After `cluster.joinSeedNodes` it's done.